### PR TITLE
[Loader][Serializer]: Add tests for Typed Function References proposal's types

### PR DESCRIPTION
**Pull request by elhewaty**

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by elhewaty** [comment]
I think the type function reference proposal was implemented but there are no test cases that cover it, so I will start by adding tests for:
- [ ] Types.
- [ ] Instructions
that were added to the proposal.


I think I am declaring `ref $t` types the wrong way, can you help with this?
another problem is that I don't understand types or instructions with negative opcodes.
> T(FuncRef, 0x70, "func")           // -0x10 for heap type

why is it encoded using `0x70` but written in negative `-0x10` in the proposal?

**Comment by elhewaty** [comment]
Oh, CC @hydai

**Comment by hydai** [comment]
@q82419 PTAL

**Comment by dannypsnl** [comment]
@elhewaty there are some commented code, do you need to convert this PR as draft?

File: test/loader/serializeTypeTest.cpp [hunk 1]
```diff
@@ -275,4 +275,74 @@ TEST(serializeTypeTest, SerializeGlobalType) {
...
+      0x06U, // Global section
+      0x04U, // Content size = 4
+      0x01U, // Vector length = 1
+      0x6FU, // ExternRef type
+      0x00U, // Const mutation
+      0x0BU  // Expression
+  };
+  EXPECT_EQ(Output, Expected);
+
+  // Why does the following test give Segmentation fault
```
Comment (q82419) [inline test/loader/serializeTypeTest.cpp:316]: It's not segmentation fault. It ran into unreachable.
You should realize the constructors of `class ValType` first.

**Review by q82419 — CHANGES_REQUESTED** [review]

**Comment by q82419** [comment]
> I think the type function reference proposal was implemented but there are no test cases that cover it, so I will start by adding tests for:
> 
> * [ ]  Types.
> * [ ]  Instructions
>   that were added to the proposal.
> 
> I think I am declaring `ref $t` types the wrong way, can you help with this? another problem is that I don't understand types or instructions with negative opcodes.
> 
> > T(FuncRef, 0x70, "func")           // -0x10 for heap type
> 
> why is it encoded using `0x70` but written in negative `-0x10` in the proposal?

Please refer to this page: https://webassembly.github.io/function-references/core/appendix/index-types.html
Because the encoding of `s7`.

**Comment by elhewaty** [comment]
@hydai @q82419 @dannypsnl 
what is wrong with this test?
```
// 4. test ref null
  GlobalType.setValType(WasmEdge::ValType(WasmEdge::TypeCode::RefNull, WasmEdge::TypeCode::ExternRef));
  Output = {};
  EXPECT_TRUE(Ser.serializeSection(createGlobalSec(GlobalType), Output));
  Expected = {
      0x06U,        // Global section
      0x05U,        // Content size = 5
      0x01U,        // Vector length = 1
      0x63U,        // RefNull type
      0x6FU,        // ExternRef heap type
      0x00U,        // Const mutation
      0x0BU         // Expression End
  };
  EXPECT_EQ(Output, Expected);
```

Error:
```
Output
    Which is: { '\x6' (6), '\x4' (4), '\x1' (1), 'o' (111, 0x6F), '\0', '\v' (11, 0xB) }
  Expected
    Which is: { '\x6' (6), '\x5' (5), '\x1' (1), 'c' (99, 0x63), 'o' (111, 0x6F), '\0', '\v' (11, 0xB) }
```

why did it ignored `RefNull` op code?

**Comment by dannypsnl** [comment]
Don't ping everyone next time. From my understanding, you should check the implementation of `lib/loader/serialize/serial_type.cpp` to figure out how it runs into a problem.

**Comment by q82419** [comment]
> @hydai @q82419 @dannypsnl what is wrong with this test?
> 
> ```
> // 4. test ref null
>   GlobalType.setValType(WasmEdge::ValType(WasmEdge::TypeCode::RefNull, WasmEdge::TypeCode::ExternRef));
>   Output = {};
>   EXPECT_TRUE(Ser.serializeSection(createGlobalSec(GlobalType), Output));
>   Expected = {
>       0x06U,        // Global section
>       0x05U,        // Content size = 5
>       0x01U,        // Vector length = 1
>       0x63U,        // RefNull type
>       0x6FU,        // ExternRef heap type
>       0x00U,        // Const mutation
>       0x0BU         // Expression End
>   };
>   EXPECT_EQ(Output, Expected);
> ```
> 
> Error:
> 
> ```
> Output
>     Which is: { '\x6' (6), '\x4' (4), '\x1' (1), 'o' (111, 0x6F), '\0', '\v' (11, 0xB) }
>   Expected
>     Which is: { '\x6' (6), '\x5' (5), '\x1' (1), 'c' (99, 0x63), 'o' (111, 0x6F), '\0', '\v' (11, 0xB) }
> ```
> 
> why did it ignored `RefNull` op code?

As mentioned in the [ValType class](https://github.com/WasmEdge/WasmEdge/blob/master/include/common/types.h#L80-L85), WasmEdge migrated the `funcref` and `externref` types into `ref null funcref` and `ref null externref` respectively.
You can check the [binary format](https://webassembly.github.io/function-references/core/binary/types.html#reference-types) of reference types in the typed function reference proposal: to supporting old spec, the `funcref` reference type is equivilant to `ref null funcref`, which has a different binary encoding.
Therefore when serializing, we output the `funcref` and `externref` briefly when got the `ref null funcref` and `ref null externref` types.

**Comment by elhewaty** [comment]
@q82419, Can you review this too?

File: test/loader/serializeTypeTest.cpp [hunk 2]
```diff
@@ -275,4 +275,75 @@ TEST(serializeTypeTest, SerializeGlobalType) {
...
+TEST(serializeTypeTest, SerializeFuncRefType) {
+  std::vector<uint8_t> Output;
+  std::vector<uint8_t> Expected;
+
+  // 5. Test serialize Function References heap types.
+  //
+  //   1. Test FuncRef heap type.
+  //   2. Test ExternRef heap type.
+  //   3. Test Ref heap type.
+  //   4. Test RefNull heap type.
```
Comment (q82419) [inline test/loader/serializeTypeTest.cpp:288]: Add 5.
Test Ref typeindex 5 heap type.

```
WasmEdge::ValType(WasmEdge::TypeCode::Ref, 5);
```

**Review by q82419 — CHANGES_REQUESTED** [review]
Almost. Please help to add the fifth test, thanks!

**Comment by elhewaty** [comment]
@q82419 PTAL.

**Review by q82419 — APPROVED** [review]

**Pull request merged**